application.cpp

```diff
@@ -440,19 +492,40 @@ void application::set_thread_priority_max() {
 }
 
 void application::exec() {
+   std::exception_ptr eptr = nullptr;
    {
-      boost::asio::io_service::work work(*io_serv);
+      boost::asio::io_service::work work(io_serv);
       (void)work;
       bool more = true;
-      while( more || io_serv->run_one() ) {
-         while( io_serv->poll_one() ) {}
-         // execute the highest priority item
-         more = pri_queue.execute_highest();
+      
+      while( more || io_serv.run_one() ) {
+         if (my->_is_quiting)
+            break;
+         try {
+            while( io_serv.poll_one() ) {}
+            // execute the highest priority item
+            more = pri_queue.execute_highest();
+         } catch(...) {
+            more = false;
+            quit();
+            eptr = std::current_exception();
+            handle_exception(eptr, "application loop");
+         }
+      }
+      
+      try {
+         pri_queue.clear(); // make sure the queue is empty
+         shutdown();        // may rethrow exceptions
+      } catch(...) {
+         if (!eptr)
+            eptr = std::current_exception();
```

Comment:
Do you need to call handle_exception here?

Reply:
No, this `try/catch` is for handling exceptions rethrown by `shutdown`, and `shutdown` already called `handle_exception` before rethrowing.